### PR TITLE
Rewrite of the NametagPlus module

### PR DESCRIPTION
**Pull request by Water261**
This rewrites the module using some new Luau features, like typechecking. It changes the config[instanceName] to config.defaults[instanceName] and just tidies up the code a little.

### CODE REVIEW

**Comment by AwesomePossum212** [comment]
Sorry for the late response! I'll take a look at this once you reopen it. Thanks for helping out!

**Comment by Water261** [comment]
It's no problem, I need to get it properly working in Studio first.

**Comment by Water261** [comment]
Alright, now it should work. The issue was that I forgot to assign the frame a parent and therefore it wouldn't show up and just caused errors.

**Comment by Water261** [comment]
When merging, can you squash and merge so I don't litter the commit history.

**Comment by Water261** [comment]
Just squashed it on my side to stop it from littering my commit history. No need to squash now.

**Comment by Water261** [comment]
Uh yeah, I'm not the best when it comes to git. This was me rebasing stuff. So fun.

**Comment by Water261** [comment]
This adds variadic support as well.

**Comment by Water261** [comment]
Alrighty, everything is finished and ready to be merged.

changeTag is now a variadic function to identify what should be changed you set the Name property inside the props table to the object you want to change.

**Comment by AwesomePossum212** [comment]
Thank you so much! I'll review and approve this ASAP (there are a lot of changes though, so it could take some time).

File: .github/README.md [hunk 1]
```diff
@@ -62,8 +64,8 @@ This example, given the default setup, will create a new nametag for every playe
 **Group + Config**
 ````
 -->>Services and modules
-local nametagSystem = require(game:GetService("ReplicatedStorage"):FindFirstChild("NametagModule")
-local nametagConfig = require(nametagSystem.Configuration)
+local nametagSystem = require(game:GetService("ReplicatedStorage"):FindFirstChild("NametagModule"))
+local nametagConfig = require(game:GetService("ReplicatedStorage"):FindFirstChild("NametagModule").Configuration)
```
Comment (AwesomePossum212) [inline .github/README.md:68]: Thanks for cleaning up the mistakes!

File: src/Configuration.lua [hunk 2]
```diff
@@ -1,107 +1,111 @@
...
+	Name = "HealthBar",
+	BackgroundTransparency = 0,
+	BorderSizePixel = 0,
+	BackgroundColor3 = config.presets["colors"]["primary"],
+	BorderColor3 = config.presets["colors"]["primary"],
+	Position = UDim2.new((1 - config.options["healthBarScale"]) / 2, 0, 0.925, 0),
+	Size = UDim2.new(config.options["healthBarScale"], 0, 0.05, 0),
+	Visible = config.options["isHealthBarVisible"],
+	AnchorPoint = config.options["healthBarAnchorPoint"],
+	ZIndex = 2,
```
Comment (AwesomePossum212) [inline src/Configuration.lua:108]: Thanks for tidying up all of the commas and leading zeroes!

File: src/init.lua [hunk 3]
```diff
@@ -0,0 +1,190 @@
...
+		local finalName = utils.waitForProperty(player, "Name")
+		if shadow_holder:FindFirstChild(finalName .. suffix) ~= nil then
+			return shadow_holder:FindFirstChild(finalName .. suffix)
+		else
+			error(player.Name .. "'s tag was not found when performing getTag.")
+		end
+	end
+end
+
+local function changeTag(tag: BillboardGui, ...): nil
```
Comment (AwesomePossum212) [inline src/init.lua:104]: much appreciated

**Review by AwesomePossum212 — APPROVED** [review]
Looks good! I'll need to test it in Studio and make a few minor changes before merging, but this is a huge step towards modernizing Nametag+.

File: src/init.lua [hunk 4]
```diff
@@ -0,0 +1,190 @@
...
+	local healthBackground = healthBar:Clone()
+
+	-- Initial setup of components
+	billboard.Enabled = true
+	healthBackground.BackgroundTransparency = 0.5
+	healthBackground.Name = "HealthBackground"
+	healthBackground.ZIndex = 1
+
+	if player ~= nil then
+		local finalName = utils.waitForProperty(player, "Name")
```
Comment (AwesomePossum212) [inline src/init.lua:37]: It looks like this function was removed from Utilities. Could you either add it back to Utilities or find a work around here? Whenever I try to call addTag in Studio this line errors.

File: src/init.lua [hunk 5]
```diff
@@ -0,0 +1,190 @@
...
+				or shadow_holder:FindFirstChild(suffix) == nil
+
+			if shadow_holder:FindFirstChild(player.Name .. suffix) then
+				return shadow_holder:FindFirstChild(player.Name .. suffix)
+			end
+		else
+			error(player.Name .. "'s tag could not be found in " .. location.Name .. ". Please create a new tag first.")
+		end
+	else
+		local finalName = utils.waitForProperty(player, "Name")
```
Comment (AwesomePossum212) [inline src/init.lua:95]: Another instance of waitForProperty.

**Review by AwesomePossum212 — CHANGES_REQUESTED** [review]
Could you either add back waitForProperty or code in a workaround? This was originally added to allow the server to give players nametags as soon as they joined, even before their name is loaded in.

**Comment by Water261** [comment]
Yeah I'll add it back, I initially removed it and VSCode didn't through any syntax errors.

**Comment by Water261** [comment]
Alright, all problems have been fixed.

**Review by AwesomePossum212 — APPROVED** [review]
Tested and it works!

**Pull request merged**